diff --git a/csrc/deep_ep.cpp b/csrc/deep_ep.cpp
index 954c9ff..a75d735 100644
--- a/csrc/deep_ep.cpp
+++ b/csrc/deep_ep.cpp
@@ -544,16 +544,19 @@ Buffer::intranode_dispatch(const torch::Tensor& x,
         topk_weights_ptr = topk_weights->data_ptr<float>();
     }
 
-    // FP8 scales checks
-    float* x_scales_ptr = nullptr;
-    int num_scales = 0, scale_token_stride = 0, scale_hidden_stride = 0;
+    // scales checks
+    uint8_t* x_scales_ptr = nullptr;
+    int num_scales = 0, scale_token_stride = 0, scale_hidden_stride = 0, scale_elem_size = 0;
     if (x_scales.has_value()) {
         EP_HOST_ASSERT(x.element_size() == 1);
-        EP_HOST_ASSERT(x_scales->scalar_type() == torch::kFloat32 or x_scales->scalar_type() == torch::kInt);
+        EP_HOST_ASSERT(x_scales->scalar_type() == torch::kUInt8 or x_scales->scalar_type() == torch::kFloat32 or
+                       x_scales->scalar_type() == torch::kInt);
         EP_HOST_ASSERT(x_scales->dim() == 2);
         EP_HOST_ASSERT(x_scales->size(0) == num_tokens);
         num_scales = x_scales->dim() == 1 ? 1 : static_cast<int>(x_scales->size(1));
-        x_scales_ptr = static_cast<float*>(x_scales->data_ptr());
+        scale_elem_size = static_cast<int>(x_scales->element_size());
+        EP_HOST_ASSERT(scale_elem_size == 1 or scale_elem_size == static_cast<int>(sizeof(float)));
+        x_scales_ptr = static_cast<uint8_t*>(x_scales->data_ptr());
         scale_token_stride = static_cast<int>(x_scales->stride(0));
         scale_hidden_stride = static_cast<int>(x_scales->stride(1));
     }
@@ -663,7 +666,7 @@ Buffer::intranode_dispatch(const torch::Tensor& x,
     // Assign pointers
     topk_idx_t* recv_topk_idx_ptr = nullptr;
     float* recv_topk_weights_ptr = nullptr;
-    float* recv_x_scales_ptr = nullptr;
+    uint8_t* recv_x_scales_ptr = nullptr;
     if (topk_idx.has_value()) {
         recv_topk_idx = torch::empty({num_recv_tokens, num_topk}, topk_idx->options());
         recv_topk_weights = torch::empty({num_recv_tokens, num_topk}, topk_weights->options());
@@ -673,7 +676,7 @@ Buffer::intranode_dispatch(const torch::Tensor& x,
     if (x_scales.has_value()) {
         recv_x_scales = x_scales->dim() == 1 ? torch::empty({num_recv_tokens}, x_scales->options())
                                              : torch::empty({num_recv_tokens, num_scales}, x_scales->options());
-        recv_x_scales_ptr = static_cast<float*>(recv_x_scales->data_ptr());
+        recv_x_scales_ptr = static_cast<uint8_t*>(recv_x_scales->data_ptr());
     }
 
     // Dispatch
@@ -686,7 +689,7 @@ Buffer::intranode_dispatch(const torch::Tensor& x,
             num_channels * num_ranks * config.num_max_nvl_chunked_recv_tokens * sizeof(int) +                     // Source index buffer
             num_channels * num_ranks * config.num_max_nvl_chunked_recv_tokens * num_topk * sizeof(topk_idx_t) +   // Top-k index buffer
             num_channels * num_ranks * config.num_max_nvl_chunked_recv_tokens * num_topk * sizeof(float) +        // Top-k weight buffer
-            num_channels * num_ranks * config.num_max_nvl_chunked_recv_tokens * sizeof(float) * num_scales        // FP8 scale buffer
+            num_channels * num_ranks * config.num_max_nvl_chunked_recv_tokens * num_scales * scale_elem_size      // Scale buffer
         <= num_nvl_bytes);
     intranode::dispatch(recv_x.data_ptr(),
                         recv_x_scales_ptr,
@@ -709,6 +712,7 @@ Buffer::intranode_dispatch(const torch::Tensor& x,
                         num_scales,
                         scale_token_stride,
                         scale_hidden_stride,
+                        scale_elem_size,
                         buffer_ptrs_gpu,
                         rank,
                         num_ranks,
@@ -1024,16 +1028,19 @@ Buffer::internode_dispatch(const torch::Tensor& x,
         topk_weights_ptr = topk_weights->data_ptr<float>();
     }
 
-    // FP8 scales checks
-    float* x_scales_ptr = nullptr;
-    int num_scales = 0, scale_token_stride = 0, scale_hidden_stride = 0;
+    // scale checks
+    uint8_t* x_scales_ptr = nullptr;
+    int num_scales = 0, scale_token_stride = 0, scale_hidden_stride = 0, scale_elem_size = 0;
     if (x_scales.has_value()) {
         EP_HOST_ASSERT(x.element_size() == 1);
-        EP_HOST_ASSERT(x_scales->scalar_type() == torch::kFloat32 or x_scales->scalar_type() == torch::kInt);
+        auto scales_type = x_scales->scalar_type();
+        EP_HOST_ASSERT(scales_type == torch::kUInt8 or scales_type == torch::kFloat32 or scales_type == torch::kInt);
         EP_HOST_ASSERT(x_scales->dim() == 2);
         EP_HOST_ASSERT(x_scales->size(0) == num_tokens);
         num_scales = x_scales->dim() == 1 ? 1 : static_cast<int>(x_scales->size(1));
-        x_scales_ptr = static_cast<float*>(x_scales->data_ptr());
+        scale_elem_size = static_cast<int>(x_scales->element_size());
+        EP_HOST_ASSERT(scale_elem_size == 1 or scale_elem_size == static_cast<int>(sizeof(float)));
+        x_scales_ptr = static_cast<uint8_t*>(x_scales->data_ptr());
         scale_token_stride = static_cast<int>(x_scales->stride(0));
         scale_hidden_stride = static_cast<int>(x_scales->stride(1));
     }
@@ -1075,6 +1082,7 @@ Buffer::internode_dispatch(const torch::Tensor& x,
                                  num_scales,
                                  num_topk,
                                  num_topk,
+                                 scale_elem_size,
                                  num_ranks,
                                  num_channels,
                                  0,
@@ -1117,6 +1125,7 @@ Buffer::internode_dispatch(const torch::Tensor& x,
                                    num_channels,
                                    hidden_int4,
                                    num_scales,
+                                   scale_elem_size,
                                    num_topk,
                                    expert_alignment,
                                    rdma_channel_prefix_matrix.data_ptr<int>(),
@@ -1186,7 +1195,7 @@ Buffer::internode_dispatch(const torch::Tensor& x,
     // Assign pointers
     topk_idx_t* recv_topk_idx_ptr = nullptr;
     float* recv_topk_weights_ptr = nullptr;
-    float* recv_x_scales_ptr = nullptr;
+    uint8_t* recv_x_scales_ptr = nullptr;
     if (topk_idx.has_value()) {
         recv_topk_idx = torch::empty({num_recv_tokens, num_topk}, topk_idx->options());
         recv_topk_weights = torch::empty({num_recv_tokens, num_topk}, topk_weights->options());
@@ -1196,7 +1205,7 @@ Buffer::internode_dispatch(const torch::Tensor& x,
     if (x_scales.has_value()) {
         recv_x_scales = x_scales->dim() == 1 ? torch::empty({num_recv_tokens}, x_scales->options())
                                              : torch::empty({num_recv_tokens, num_scales}, x_scales->options());
-        recv_x_scales_ptr = static_cast<float*>(recv_x_scales->data_ptr());
+        recv_x_scales_ptr = static_cast<uint8_t*>(recv_x_scales->data_ptr());
     }
 
     // Launch data dispatch
@@ -1227,6 +1236,7 @@ Buffer::internode_dispatch(const torch::Tensor& x,
                         num_experts,
                         scale_token_stride,
                         scale_hidden_stride,
+                        scale_elem_size,
                         rdma_buffer_ptr,
                         config.num_max_rdma_chunked_send_tokens,
                         config.num_max_rdma_chunked_recv_tokens,
@@ -1394,6 +1404,7 @@ std::tuple<torch::Tensor, std::optional<torch::Tensor>, std::optional<EventHandl
                              0,
                              0,
                              num_topk,
+                             0,
                              num_ranks,
                              num_channels,
                              num_combined_tokens,
diff --git a/csrc/kernels/api.cuh b/csrc/kernels/api.cuh
index 9bbe096..41e7988 100644
--- a/csrc/kernels/api.cuh
+++ b/csrc/kernels/api.cuh
@@ -76,14 +76,14 @@ void cached_notify_dispatch(const int* rank_prefix_matrix,
                             cudaStream_t stream);
 
 void dispatch(void* recv_x,
-              float* recv_x_scales,
+              uint8_t* recv_x_scales,
               int* recv_src_idx,
               topk_idx_t* recv_topk_idx,
               float* recv_topk_weights,
               int* recv_channel_offset,
               int* send_head,
               const void* x,
-              const float* x_scales,
+              const uint8_t* x_scales,
               const topk_idx_t* topk_idx,
               const float* topk_weights,
               const bool* is_token_in_rank,
@@ -96,6 +96,7 @@ void dispatch(void* recv_x,
               int num_scales,
               int scale_token_stride,
               int scale_hidden_stride,
+              int scale_elem_size,
               void** buffer_ptrs,
               int rank,
               int num_ranks,
@@ -158,6 +159,7 @@ void notify_dispatch(const int* num_tokens_per_rank,
                      int num_channels,
                      int hidden_int4,
                      int num_scales,
+                     int scale_elem_size,
                      int num_topk,
                      int expert_alignment,
                      int* rdma_channel_prefix_matrix,
@@ -176,12 +178,12 @@ void notify_dispatch(const int* num_tokens_per_rank,
                      bool low_latency_mode);
 
 void dispatch(void* recv_x,
-              float* recv_x_scales,
+              uint8_t* recv_x_scales,
               topk_idx_t* recv_topk_idx,
               float* recv_topk_weights,
               void* recv_src_meta,
               const void* x,
-              const float* x_scales,
+              const uint8_t* x_scales,
               const topk_idx_t* topk_idx,
               const float* topk_weights,
               int* send_rdma_head,
@@ -201,6 +203,7 @@ void dispatch(void* recv_x,
               int num_experts,
               int scale_token_stride,
               int scale_hidden_stride,
+              int scale_elem_size,
               void* rdma_buffer_ptr,
               int num_max_rdma_chunked_send_tokens,
               int num_max_rdma_chunked_recv_tokens,
@@ -218,6 +221,7 @@ void cached_notify(int hidden_int4,
                    int num_scales,
                    int num_topk_idx,
                    int num_topk_weights,
+                   int scale_elem_size,
                    int num_ranks,
                    int num_channels,
                    int num_combined_tokens,
diff --git a/csrc/kernels/internode.cu b/csrc/kernels/internode.cu
index 48c6c00..2db0683 100644
--- a/csrc/kernels/internode.cu
+++ b/csrc/kernels/internode.cu
@@ -39,8 +39,9 @@ int get_source_meta_bytes() {
     return sizeof(SourceMeta);
 }
 
-__host__ __device__ __forceinline__ int get_num_bytes_per_token(int hidden_int4, int num_scales, int num_topk_idx, int num_topk_weights) {
-    return static_cast<int>(align_up(hidden_int4 * sizeof(int4) + sizeof(SourceMeta) + num_scales * sizeof(float) +
+__host__ __device__ __forceinline__ int get_num_bytes_per_token(int hidden_int4, int num_scales, int num_topk_idx, int num_topk_weights, int scale_elem_size) {
+    auto scale_bytes = static_cast<int64_t>(num_scales) * (scale_elem_size == 0 ? sizeof(uint8_t) : scale_elem_size);
+    return static_cast<int>(align_up(hidden_int4 * sizeof(int4) + sizeof(SourceMeta) + scale_bytes +
                                          num_topk_idx * sizeof(int) + num_topk_weights * sizeof(float),
                                      sizeof(int4)));
 }
@@ -49,11 +50,12 @@ __host__ __device__ __forceinline__ std::pair<int, int> get_rdma_clean_meta(int
                                                                             int num_scales,
                                                                             int num_topk_idx,
                                                                             int num_topk_weights,
+                                                                            int scale_elem_size,
                                                                             int num_rdma_ranks,
                                                                             int num_rdma_recv_buffer_tokens,
                                                                             int num_channels) {
     // Return `int32_t` offset and count to clean
-    return {(get_num_bytes_per_token(hidden_int4, num_scales, num_topk_idx, num_topk_weights) * num_rdma_recv_buffer_tokens *
+    return {(get_num_bytes_per_token(hidden_int4, num_scales, num_topk_idx, num_topk_weights, scale_elem_size) * num_rdma_recv_buffer_tokens *
              num_rdma_ranks * 2 * num_channels) /
                 sizeof(int),
             (NUM_MAX_NVL_PEERS * 2 + 4) * num_rdma_ranks * 2 * num_channels};
@@ -63,6 +65,7 @@ __host__ __device__ __forceinline__ std::pair<int, int> get_nvl_clean_meta(int h
                                                                            int num_scales,
                                                                            int num_topk_idx,
                                                                            int num_topk_weights,
+                                                                           int scale_elem_size,
                                                                            int num_rdma_ranks,
                                                                            int num_nvl_ranks,
                                                                            int num_nvl_recv_buffer_tokens,
@@ -72,9 +75,9 @@ __host__ __device__ __forceinline__ std::pair<int, int> get_nvl_clean_meta(int h
     EP_STATIC_ASSERT(sizeof(SourceMeta) % sizeof(int) == 0, "Invalid size of `SourceMeta`");
 
     return {
-        (num_nvl_recv_buffer_tokens * get_num_bytes_per_token(hidden_int4, num_scales, num_topk_idx, num_topk_weights) * num_nvl_ranks *
-         num_channels) /
-            sizeof(int),
+        (num_nvl_recv_buffer_tokens * get_num_bytes_per_token(hidden_int4, num_scales, num_topk_idx, num_topk_weights, scale_elem_size) * num_nvl_ranks *
+             num_channels) /
+                sizeof(int),
         num_nvl_ranks * (2 * num_rdma_ranks + 2) * num_channels,
     };
 }
@@ -357,6 +360,7 @@ void notify_dispatch(const int* num_tokens_per_rank,
                      int num_channels,
                      int hidden_int4,
                      int num_scales,
+                     int scale_elem_size,
                      int num_topk,
                      int expert_alignment,
                      int* rdma_channel_prefix_matrix,
@@ -411,12 +415,13 @@ void notify_dispatch(const int* num_tokens_per_rank,
     const auto num_rdma_ranks = num_ranks / NUM_MAX_NVL_PEERS;
 
     // Get clean meta
-    auto rdma_clean_meta =
-        get_rdma_clean_meta(hidden_int4, num_scales, num_topk, num_topk, num_rdma_ranks, num_max_rdma_chunked_recv_tokens, num_channels);
+    auto rdma_clean_meta = 
+        get_rdma_clean_meta(hidden_int4, num_scales, num_topk, num_topk, scale_elem_size, num_rdma_ranks, num_max_rdma_chunked_recv_tokens, num_channels);
     auto nvl_clean_meta = get_nvl_clean_meta(hidden_int4,
                                              num_scales,
                                              num_topk,
                                              num_topk,
+                                             scale_elem_size,
                                              num_rdma_ranks,
                                              NUM_MAX_NVL_PEERS,
                                              num_max_nvl_chunked_recv_tokens,
@@ -446,12 +451,12 @@ template <bool kLowLatencyMode,
           int kNumTopkRDMARanks = get_num_topk_rdma_ranks(kNumRDMARanks)>
 __global__ void __launch_bounds__(((kNumDispatchRDMASenderWarps + 1 + NUM_MAX_NVL_PEERS) * 32), 1)
     dispatch(int4* recv_x,
-             float* recv_x_scales,
+             uint8_t* recv_x_scales,
              topk_idx_t* recv_topk_idx,
              float* recv_topk_weights,
              SourceMeta* recv_src_meta,
              const int4* x,
-             const float* x_scales,
+             const uint8_t* x_scales,
              const topk_idx_t* topk_idx,
              const float* topk_weights,
              int* send_rdma_head,
@@ -471,6 +476,7 @@ __global__ void __launch_bounds__(((kNumDispatchRDMASenderWarps + 1 + NUM_MAX_NV
              int num_experts,
              int scale_token_stride,
              int scale_hidden_stride,
+             int scale_elem_size,
              void* rdma_buffer_ptr,
              int num_max_rdma_chunked_send_tokens,
              int num_max_rdma_chunked_recv_tokens,
@@ -516,8 +522,10 @@ __global__ void __launch_bounds__(((kNumDispatchRDMASenderWarps + 1 + NUM_MAX_NV
     // RDMA symmetric layout
     EP_STATIC_ASSERT(NUM_MAX_NVL_PEERS * sizeof(bool) == sizeof(uint64_t), "Invalid number of NVL peers");
     auto hidden_bytes = hidden_int4 * sizeof(int4);
-    auto scale_bytes = num_scales * sizeof(float);
-    auto num_bytes_per_token = get_num_bytes_per_token(hidden_int4, num_scales, num_topk, num_topk);
+    auto scale_elem_bytes = scale_elem_size == 0 ? sizeof(uint8_t) : scale_elem_size;
+    bool scale_is_fp32 = scale_elem_size == sizeof(float);
+    auto scale_bytes = num_scales * scale_elem_bytes;
+    auto num_bytes_per_token = get_num_bytes_per_token(hidden_int4, num_scales, num_topk, num_topk, scale_elem_size);
     auto rdma_channel_data = SymBuffer<uint8_t>(
         rdma_buffer_ptr, num_max_rdma_chunked_recv_tokens * num_bytes_per_token, kNumRDMARanks, channel_id, num_channels);
     auto rdma_channel_meta = SymBuffer<int>(rdma_buffer_ptr, NUM_MAX_NVL_PEERS * 2 + 2, kNumRDMARanks, channel_id, num_channels);
@@ -692,17 +700,31 @@ __global__ void __launch_bounds__(((kNumDispatchRDMASenderWarps + 1 + NUM_MAX_NV
                 dst_send_buffers[i] = reinterpret_cast<int4*>(dst_send_buffers[i]) + hidden_int4;
 
             // Copy `x_scales` into symmetric send buffer
-            #pragma unroll
-            for (int i = lane_id; i < num_scales; i += 32) {
-                auto offset = token_idx * scale_token_stride + i * scale_hidden_stride;
-                auto value = ld_nc_global(x_scales + offset);
+            if (num_scales > 0) {
+                if (scale_is_fp32) {
+                    auto x_scales_fp32 = reinterpret_cast<const float*>(x_scales);
+                    #pragma unroll
+                    for (int i = lane_id; i < num_scales; i += 32) {
+                        auto offset = token_idx * scale_token_stride + i * scale_hidden_stride;
+                        auto value = __ldg(x_scales_fp32 + offset);
+                        #pragma unroll
+                        for (int j = 0; j < num_topk_ranks; ++j)
+                            st_na_global(reinterpret_cast<float*>(dst_send_buffers[j]) + i, value);
+                    }
+                } else {
+                    #pragma unroll
+                    for (int i = lane_id; i < num_scales; i += 32) {
+                        auto offset = token_idx * scale_token_stride + i * scale_hidden_stride;
+                        auto value = ld_nc_global(x_scales + offset);
+                        #pragma unroll
+                        for (int j = 0; j < num_topk_ranks; ++j)
+                            st_na_global(reinterpret_cast<uint8_t*>(dst_send_buffers[j]) + i, value);
+                    }
+                }
                 #pragma unroll
-                for (int j = 0; j < num_topk_ranks; ++j)
-                    st_na_global(reinterpret_cast<float*>(dst_send_buffers[j]) + i, value);
+                for (int i = 0; i < num_topk_ranks; ++i)
+                    dst_send_buffers[i] = reinterpret_cast<uint8_t*>(dst_send_buffers[i]) + scale_bytes;
             }
-            #pragma unroll
-            for (int i = 0; i < num_topk_ranks; ++i)
-                dst_send_buffers[i] = reinterpret_cast<float*>(dst_send_buffers[i]) + num_scales;
 
             // Copy source metadata into symmetric send buffer
             if (lane_id < num_topk_ranks)
@@ -1141,20 +1163,20 @@ __global__ void __launch_bounds__(((kNumDispatchRDMASenderWarps + 1 + NUM_MAX_NV
                 mbarrier_wait(tma_mbarrier, tma_phase);
                 if (elect_one_sync()) {
                     tma_store_1d(tma_buffer, recv_x + recv_token_idx * hidden_int4, hidden_bytes, false);
-                    if (scale_aligned)
-                        tma_store_1d(tma_buffer + hidden_bytes, recv_x_scales + recv_token_idx * num_scales, scale_bytes, false);
+                    if (scale_aligned and scale_bytes > 0)
+                        tma_store_1d(tma_buffer + hidden_bytes, recv_x_scales + static_cast<int64_t>(recv_token_idx) * scale_bytes, scale_bytes, false);
                 }
                 __syncwarp();
                 shifted += hidden_bytes;
 
                 // Copy scales
                 // TODO: make it as templated
-                if (not scale_aligned) {
+                if (not scale_aligned and scale_bytes > 0) {
                     UNROLLED_WARP_COPY(1,
                                        lane_id,
-                                       num_scales,
-                                       recv_x_scales + recv_token_idx * num_scales,
-                                       reinterpret_cast<float*>(shifted),
+                                       scale_bytes,
+                                       recv_x_scales + static_cast<int64_t>(recv_token_idx) * scale_bytes,
+                                       reinterpret_cast<uint8_t*>(shifted),
                                        ld_nc_global,
                                        st_na_global);
                 }
@@ -1208,12 +1230,12 @@ __global__ void __launch_bounds__(((kNumDispatchRDMASenderWarps + 1 + NUM_MAX_NV
 }
 
 void dispatch(void* recv_x,
-              float* recv_x_scales,
+              uint8_t* recv_x_scales,
               topk_idx_t* recv_topk_idx,
               float* recv_topk_weights,
               void* recv_src_meta,
               const void* x,
-              const float* x_scales,
+              const uint8_t* x_scales,
               const topk_idx_t* topk_idx,
               const float* topk_weights,
               int* send_rdma_head,
@@ -1233,6 +1255,7 @@ void dispatch(void* recv_x,
               int num_experts,
               int scale_token_stride,
               int scale_hidden_stride,
+              int scale_elem_size,
               void* rdma_buffer_ptr,
               int num_max_rdma_chunked_send_tokens,
               int num_max_rdma_chunked_recv_tokens,
@@ -1288,6 +1311,7 @@ void dispatch(void* recv_x,
                       num_experts,                                                                                             \
                       scale_token_stride,                                                                                      \
                       scale_hidden_stride,                                                                                     \
+                      scale_elem_size,                                                                                         \
                       rdma_buffer_ptr,                                                                                         \
                       num_max_rdma_chunked_send_tokens,                                                                        \
                       num_max_rdma_chunked_recv_tokens,                                                                        \
@@ -1471,6 +1495,7 @@ void cached_notify(int hidden_int4,
                    int num_scales,
                    int num_topk_idx,
                    int num_topk_weights,
+                   int scale_elem_size,
                    int num_ranks,
                    int num_channels,
                    int num_combined_tokens,
@@ -1497,11 +1522,12 @@ void cached_notify(int hidden_int4,
 
     // Get clean meta
     auto rdma_clean_meta = get_rdma_clean_meta(
-        hidden_int4, num_scales, num_topk_idx, num_topk_weights, num_rdma_ranks, num_max_rdma_chunked_recv_tokens, num_channels);
+        hidden_int4, num_scales, num_topk_idx, num_topk_weights, scale_elem_size, num_rdma_ranks, num_max_rdma_chunked_recv_tokens, num_channels);
     auto nvl_clean_meta = get_nvl_clean_meta(hidden_int4,
                                              num_scales,
                                              num_topk_idx,
                                              num_topk_weights,
+                                             scale_elem_size,
                                              num_rdma_ranks,
                                              NUM_MAX_NVL_PEERS,
                                              num_max_nvl_chunked_recv_tokens,
@@ -1750,7 +1776,7 @@ __global__ void __launch_bounds__((kNumForwarders + 1) * 32, 1) combine(int4* co
     EP_DEVICE_ASSERT(hidden % (sizeof(int4) / sizeof(dtype_t)) == 0);
     const auto hidden_int4 = hidden / (sizeof(int4) / sizeof(dtype_t));
     const auto hidden_bytes = hidden_int4 * sizeof(int4);
-    const auto num_bytes_per_token = get_num_bytes_per_token(hidden_int4, 0, 0, num_topk);
+    const auto num_bytes_per_token = get_num_bytes_per_token(hidden_int4, 0, 0, num_topk, 0);
 
     // NOTES: we decouple a channel into 2 SMs
     const auto rdma_rank = rank / NUM_MAX_NVL_PEERS, nvl_rank = rank % NUM_MAX_NVL_PEERS;
diff --git a/csrc/kernels/intranode.cu b/csrc/kernels/intranode.cu
index 44b0cef..f10b4fc 100644
--- a/csrc/kernels/intranode.cu
+++ b/csrc/kernels/intranode.cu
@@ -196,14 +196,14 @@ void cached_notify_dispatch(const int* rank_prefix_matrix,
 
 template <int kNumRanks, int kNumThreads, int kNumTMABytesPerWarp>
 __global__ void __launch_bounds__(kNumThreads, 1) dispatch(int4* recv_x,
-                                                           float* recv_x_scales,
+                                                           uint8_t* recv_x_scales,
                                                            int* recv_src_idx,
                                                            topk_idx_t* recv_topk_idx,
                                                            float* recv_topk_weights,
                                                            int* recv_channel_offset,
                                                            int* send_head,
                                                            const int4* x,
-                                                           const float* x_scales,
+                                                           const uint8_t* x_scales,
                                                            const topk_idx_t* topk_idx,
                                                            const float* topk_weights,
                                                            const bool* is_token_in_rank,
@@ -216,6 +216,7 @@ __global__ void __launch_bounds__(kNumThreads, 1) dispatch(int4* recv_x,
                                                            int num_scales,
                                                            int scale_token_stride,
                                                            int scale_hidden_stride,
+                                                           int scale_elem_size,
                                                            void** buffer_ptrs,
                                                            int rank,
                                                            int num_max_send_tokens,
@@ -264,7 +265,7 @@ __global__ void __launch_bounds__(kNumThreads, 1) dispatch(int4* recv_x,
     // `src_idx_buffers`: kNumChannels * kNumRanks * num_recv_buffer_tokens * sizeof(int)
     // `topk_idx_buffers`: kNumChannels * kNumRanks * num_recv_buffer_tokens * num_topk * sizeof(topk_idx_t)
     // `topk_weights_buffers`: kNumChannels * kNumRanks * num_recv_buffer_tokens * num_topk * sizeof(float)
-    // `x_scales_buffers`: kNumChannels * kNumRanks * num_recv_buffer_tokens * num_scales * sizeof(float)
+    // `x_scales_buffers`: kNumChannels * kNumRanks * num_recv_buffer_tokens * num_scales * scale_elem_size bytes
     auto channel_x_buffers = Buffer<int4>(
         ptr, num_channels_total * num_recv_buffer_tokens * hidden_int4, channel_rank_offset * num_recv_buffer_tokens * hidden_int4);
     auto channel_src_idx_buffers =
@@ -273,8 +274,13 @@ __global__ void __launch_bounds__(kNumThreads, 1) dispatch(int4* recv_x,
         ptr, num_channels_total * num_recv_buffer_tokens * num_topk, channel_rank_offset * num_recv_buffer_tokens * num_topk);
     auto channel_topk_weights_buffers =
         Buffer<float>(ptr, num_channels_total * num_recv_buffer_tokens * num_topk, channel_rank_offset * num_recv_buffer_tokens * num_topk);
-    auto channel_x_scales_buffers = Buffer<float>(
-        ptr, num_channels_total * num_recv_buffer_tokens * num_scales, channel_rank_offset * num_recv_buffer_tokens * num_scales);
+    int scale_elem_bytes = scale_elem_size == 0 ? 1 : scale_elem_size;
+    EP_DEVICE_ASSERT(scale_elem_size == 0 or scale_elem_size == 1 or scale_elem_size == sizeof(float));
+    bool scale_is_fp32 = scale_elem_size == sizeof(float);
+    int scale_bytes = num_scales * scale_elem_bytes;
+    auto channel_x_scales_buffers = Buffer<uint8_t>(ptr,
+                                                    num_channels_total * num_recv_buffer_tokens * scale_bytes,
+                                                    channel_rank_offset * num_recv_buffer_tokens * scale_bytes);
 
     // TMA stuffs
 #ifndef DISABLE_SM90_FEATURES
@@ -379,10 +385,22 @@ __global__ void __launch_bounds__(kNumThreads, 1) dispatch(int4* recv_x,
                     }
 
                     // Copy `x_scales`
-                    #pragma unroll
-                    for (int i = lane_id; i < num_scales; i += 32) {
-                        auto offset = token_idx * scale_token_stride + i * scale_hidden_stride;
-                        channel_x_scales_buffers[dst_slot_idx * num_scales + i] = __ldg(x_scales + offset);
+                    if (num_scales > 0) {
+                        if (scale_is_fp32) {
+                            auto x_scales_fp32 = reinterpret_cast<const float*>(x_scales);
+                            auto channel_scales_fp32 = reinterpret_cast<float*>(channel_x_scales_buffers.buffer());
+                            #pragma unroll
+                            for (int i = lane_id; i < num_scales; i += 32) {
+                                auto offset = token_idx * scale_token_stride + i * scale_hidden_stride;
+                                channel_scales_fp32[dst_slot_idx * num_scales + i] = __ldg(x_scales_fp32 + offset);
+                            }
+                        } else {
+                            #pragma unroll
+                            for (int i = lane_id; i < num_scales; i += 32) {
+                                auto offset = token_idx * scale_token_stride + i * scale_hidden_stride;
+                                channel_x_scales_buffers[dst_slot_idx * num_scales + i] = __ldg(x_scales + offset);
+                            }
+                        }
                     }
                 }
 
@@ -498,12 +516,15 @@ __global__ void __launch_bounds__(kNumThreads, 1) dispatch(int4* recv_x,
             }
 
             // Copy `x_scales`
-            #pragma unroll 4
-            for (int i = recv_thread_id_in_rank; i < num_recv_tokens * num_scales; i += 32 * num_recv_warps_per_rank) {
-                int chunk_idx = i / num_scales, scales_idx = i % num_scales;
-                int token_idx_in_buffer = (cached_channel_head_idx + chunk_idx) % num_recv_buffer_tokens;
-                recv_x_scales[static_cast<int64_t>(total_offset + chunk_idx) * num_scales + scales_idx] =
-                    ld_nc_global(channel_x_scales_buffers.buffer() + token_idx_in_buffer * num_scales + scales_idx);
+            if (scale_bytes > 0) {
+                #pragma unroll 4
+                for (int i = recv_thread_id_in_rank; i < num_recv_tokens * scale_bytes; i += 32 * num_recv_warps_per_rank) {
+                    int chunk_idx = i / scale_bytes;
+                    int byte_idx = i - chunk_idx * scale_bytes;
+                    int token_idx_in_buffer = (cached_channel_head_idx + chunk_idx) % num_recv_buffer_tokens;
+                    recv_x_scales[static_cast<int64_t>(total_offset + chunk_idx) * scale_bytes + byte_idx] =
+                        ld_nc_global(channel_x_scales_buffers.buffer() + token_idx_in_buffer * scale_bytes + byte_idx);
+                }
             }
 
             // Move queue
@@ -532,14 +553,14 @@ __global__ void __launch_bounds__(kNumThreads, 1) dispatch(int4* recv_x,
 }
 
 void dispatch(void* recv_x,
-              float* recv_x_scales,
+              uint8_t* recv_x_scales,
               int* recv_src_idx,
               topk_idx_t* recv_topk_idx,
               float* recv_topk_weights,
               int* recv_channel_offset,
               int* send_head,
               const void* x,
-              const float* x_scales,
+              const uint8_t* x_scales,
               const topk_idx_t* topk_idx,
               const float* topk_weights,
               const bool* is_token_in_rank,
@@ -552,6 +573,7 @@ void dispatch(void* recv_x,
               int num_scales,
               int scale_token_stride,
               int scale_hidden_stride,
+              int scale_elem_size,
               void** buffer_ptrs,
               int rank,
               int num_ranks,
@@ -595,6 +617,7 @@ void dispatch(void* recv_x,
                       num_scales,                                        \
                       scale_token_stride,                                \
                       scale_hidden_stride,                               \
+                      scale_elem_size,                                   \
                       buffer_ptrs,                                       \
                       rank,                                              \
                       num_max_send_tokens,                               \
diff --git a/deep_ep/buffer.py b/deep_ep/buffer.py
index 37512ee..e769e37 100644
--- a/deep_ep/buffer.py
+++ b/deep_ep/buffer.py
@@ -339,8 +339,8 @@ class Buffer:
         Arguments:
             x: `torch.Tensor` or tuple of `torch.Tensor`, for the first type, the shape must be `[num_tokens, hidden]`,
                 and type must be `torch.bfloat16`; for the second type, the first element of the tuple must be shaped as
-                `[num_tokens, hidden]` with type `torch.float8_e4m3fn`, the second must be `[num_tokens, hidden // 128]`
-                 (requiring divisible) with type `torch.float`.
+                `[num_tokens, hidden]` with type `torch.uint8`, the second must be `[num_tokens, hidden // 128 or 16]`
+                 (requiring divisible) with type `torch.uint8` (MXFP4) or `torch.float32` (FP8 with per-128 scales).
             handle: an optional communication handle, if set, the CPU will reuse the layout information to save some time.
             num_tokens_per_rank: `[num_ranks]` with `torch.int`, the number of tokens to be sent to each rank.
             num_tokens_per_rdma_rank: `[num_rdma_ranks]` with `torch.int`, the number of tokens to be sent to each RDMA
